lib/IRGen/GenMeta.cpp

```diff
@@ -5054,13 +5057,88 @@ static void emitObjCClassSymbol(IRGenModule &IGM, ClassDecl *classDecl,
       .to(alias, link.isForDefinition());
 }
 
+/// Check whether the metadata update strategy requires runtime support that is
+/// not guaranteed available by the minimum deployment target, and diagnose if
+/// so.
+static void
+diagnoseUnsupportedObjCImplLayout(IRGenModule &IGM, ClassDecl *classDecl,
+                                  const ClassLayout &fragileLayout) {
+  if (!fragileLayout.hasObjCImplementation())
+    return;
+
+  auto strategy = IGM.getClassMetadataStrategy(classDecl);
+
+  switch (strategy) {
+  case ClassMetadataStrategy::Fixed:
+    // Fixed is just fine; no special support needed.
+    break;
+
+  case ClassMetadataStrategy::FixedOrUpdate:
+  case ClassMetadataStrategy::Update: {
+    auto &ctx = IGM.Context;
+
+    // Update and FixedOrUpdate require support in both the Swift and ObjC
+    // runtimes.
+    auto requiredAvailability=ctx.getUpdatePureObjCClassMetadataAvailability();
+    // FIXME: Take the class's availability into account
```

Comment:
Out of curiosity, does class metadata layout happen lazily when the first instance is created or something? I'm wondering how gating this on the availability of the class itself would be safe in practice.

Reply:
The first time the ObjC runtime tries to access any of the metadata associated with a class—basically, use it as anything *but* a completely opaque pointer—it performs an operation called "realization". The metadata update function that Swift generates will be called during realization, and it will call the new `swift_updatePureObjCClassMetadata()` function. So realization might happen when you try to look up the `+alloc` method to allocate an instance of the class, or it could happen even earlier, depending on how the code uses the class.